### PR TITLE
make DType a dataclass

### PR DESCRIPTION
**Pull request by thelonejordan**
- [x] Make `DType` a dataclass

* directly using `DType` object as keys creates issues while subclassing the dataclass, so used the `name` attribute as key instead.
* added support for `PtrDType` to `DType` comparison. (this was not needed for `NamedTuple`).

### CODE REVIEW

**Comment by chenyuxyz** [comment]
making DType a dataclass is good, but i don't see the point of making np a property. the goal is to remove .np!

**Comment by thelonejordan** [comment]
> making DType a dataclass is good, but i don't see the point of making np a property. the goal is to remove .np!

@chenyuxyz the goal was to remove numpy incrementally. Removing numpy is a big change in itself https://github.com/tinygrad/tinygrad/pull/3075. I was trying to limit this PR to just converting NamedTuple to dataclass. We can always delete later.

**Comment by chenyuxyz** [comment]
is making .np a property a requirement to make DType a dataclass?

**Comment by thelonejordan** [comment]
> is making .np a property a requirement to make DType a dataclass?

@chenyuxyz It's not. There was a TODO to remove np from attributes. So I removed it and added "np" as a property to support the .np calls. The functionality is exactly the same, the numpy to dtype mapping is handled dynamically, so cleaner.

File: tinygrad/dtype.py [hunk 1]
```diff
@@ -70,9 +73,9 @@ def fields() -> Dict[str, DType]: return DTYPES_DICT
 
   # NOTE: these are image dtypes
   @staticmethod
-  def imageh(shp): return ImageDType(100, 2, "imageh", np.float16, shp, dtypes.float32)
```
Comment (thelonejordan) [inline tinygrad/dtype.py:73]: this change is ambiguous!

Comment (thelonejordan) [reply tinygrad/dtype.py:73]: fixed!

**Comment by thelonejordan** [comment]
> is making .np a property a requirement to make DType a dataclass?

@chenyuxyz reverted this.

**Comment by geohot** [comment]
Cool!

**Pull request merged**